go-selinux/selinux_linux.go

```diff
@@ -445,6 +469,194 @@ func ComputeCreateContext(source string, target string, class string) (string, e
+			catstart, err := parseLevelItem(ranges[0], category)
+			if err != nil {
+				return nil, err
+			}
+			catend, err := parseLevelItem(ranges[1], category)
+			if err != nil {
+				return nil, err
+			}
+			for i := catstart; i <= catend; i++ {
+				bitset.Set(uint(i))
+			}
+		} else {
+			cat, err := parseLevelItem(ranges[0], category)
+			if err != nil {
+				return nil, err
+			}
+			bitset.Set(uint(cat))
+		}
+	}
+
+	return bitset, nil
+}
+
+// parseLevelItem parses and verifies that a sensitivity or category are valid
+func parseLevelItem(s string, sep levelItem) (uint32, error) {
+	item := []rune(s)
+	if item[0] != rune(sep) {
+		return 0, ErrLevelSyntax
+	}
+	val, err := strconv.ParseUint(string(item)[1:], 10, 64)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to parse level")
+	}
+
+	return uint32(val), nil
+}
+
+// rangeStrToMLSRange marshals a string representation of a range.
+func rangeStrToMLSRange(rangeStr string) (*mlsRange, error) {
+	mlsRange := &mlsRange{}
+	levelSlice := strings.SplitN(rangeStr, "-", 2)
+
+	switch len(levelSlice) {
+	case 2:
+		high := strings.SplitN(levelSlice[1], ":", 2)
+		sensHigh, err := parseLevelItem(high[0], sensitivity)
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.high = &level{sens: sensHigh}
+		if len(high) > 1 {
+			catsHigh, err := catsToBitset(high[1])
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse categories")
+			}
+			mlsRange.high.cats = catsHigh
+		}
+		fallthrough
+	case 1:
+		low := strings.SplitN(levelSlice[0], ":", 2)
+		sensLow, err := parseLevelItem(low[0], sensitivity)
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.low = &level{sens: sensLow}
+		if len(low) > 1 {
+			catsLow, err := catsToBitset(low[1])
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse categories")
+			}
+			mlsRange.low.cats = catsLow
+		}
+	}
+
+	if mlsRange.high == nil {
+		mlsRange.high = mlsRange.low
+	}
+
+	return mlsRange, nil
+}
+
+/*
+bitsetToStr takes a category bitset and returns it in the
+canonical selinux syntax
+*/
+func bitsetToStr(c *bitset.BitSet) string {
+	var catBuff bytes.Buffer
+	i, e := c.NextSet(0)
+	len := 0
+	for e {
+		if len == 0 {
+			if catBuff.Len() > 0 {
+				catBuff.WriteString(",")
+			}
+			catBuff.WriteString("c" + strconv.FormatInt(int64(i), 10))
+		}
+
+		next, e := c.NextSet(i + 1)
+		if e {
+			// consecutive cats
+			if next == i+1 {
+				len++
+				i = next
+				continue
+			}
+		}
+		if len == 1 {
+			catBuff.WriteString(",c" + strconv.FormatInt(int64(i), 10))
+		} else if len > 1 {
+			catBuff.WriteString(".c" + strconv.FormatInt(int64(i), 10))
+		}
+		if !e {
+			break
+		}
+		len = 0
+		i = next
+	}
+
+	return catBuff.String()
+}
+
+// String returns an mlsRange as a string.
+func (m mlsRange) String() string {
+	low := fmt.Sprintf("s%d", m.low.sens)
+	if m.low.cats != nil && m.low.cats.Count() > 0 {
+		low += ":" + bitsetToStr(m.low.cats)
+	}
+
+	if reflect.DeepEqual(m.low, m.high) {
```

Comment:
This (`reflect.DeepEqual`) is probably too costly to use anywhere expect for tests (or when you don't know the data format you're dealing with). Please write a comparison function.

Reply:
will do